File: 1509-minimum-difference-between-largest-and-smallest-value-in-three-moves/1509-minimum-difference-between-largest-and-smallest-value-in-three-moves.cpp
class Solution {
public:
    int minDifference(vector<int>& nums) {
        int n = nums.size(), minDiff = INT_MAX;
        if (n <= 4) return 0;
        
        // Partially sort the first four elements
        partial_sort(nums.begin(), nums.begin() + 4, nums.end());
        
        // Find the 4th largest element
        nth_element(nums.begin() + 4, nums.begin() + (n - 4), nums.end());
        
        // Sort the last four elements
        sort(nums.begin() + (n - 4), nums.end());
        
        // Four scenarios to compute the minimum difference
        for (int left = 0, right = n - 4; left < 4; ++left, ++right)
            minDiff = min(minDiff, nums[right] - nums[left]);
        
        return minDiff;
    }
};

File: 1509-minimum-difference-between-largest-and-smallest-value-in-three-moves/1509-minimum-difference-between-largest-and-smallest-value-in-three-moves.py
class Solution:
    def minDifference(self, nums: List[int]) -> int:
        n = len(nums)
        
        # If the array hhas 4 or fewer elements, return 0
        if n <= 4:
            return 0
        
        nums.sort()
        min_diff = float("inf")
        
        # Four scenarios to compute the minimum difference
        for left in range(4):
            right = n - 4 + left
            min_diff = min(min_diff, nums[right] - nums[left])
        
        return min_diff